File: CulinarySiteApp/src/app/services/subscriber.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ISubscriberDetailModel } from '../interfaces/subscriber/subscriber-detail-model.interface';
import { ISubscriberListModel } from '../interfaces/subscriber/subscriber-list-model.interface';
import { CreateSubscriberModel } from '../viewmodels/subscriber/create-subscriber-model.class';
import { UpdateSubscriberModel } from '../viewmodels/subscriber/update-subscriber-model.class';

@Injectable()
export class SubscriberService {
  public url: string = '/api/subscriber';

  constructor(private http: HttpClient) {}

  public getSubscriberList(
    withRelated: boolean
  ): Observable<ISubscriberListModel[]> {
    return this.http.get<ISubscriberListModel[]>(this.url + '/' + withRelated);
  }

  public getSubscriber(
    id: number,
    withRelated: boolean
  ): Observable<ISubscriberDetailModel> {
    return this.http.get<ISubscriberDetailModel>(
      this.url + '/' + id + '/' + withRelated
    );
  }

  public createSubscriber(
    createSubscriberModel: CreateSubscriberModel
  ): Observable<void> {
    return this.http.post<void>(this.url, createSubscriberModel);
  }

  public updateSubscriber(
    updateSubscriberModel: UpdateSubscriberModel
  ): Observable<void> {
    return this.http.put<void>(this.url, updateSubscriberModel);
  }

  public deleteSubscriber(id: number): Observable<void> {
    return this.http.delete<void>(this.url + '/' + id);
  }
}

File: CulinarySite/CulinarySite.Bll/Services/SubscriberService.cs
using System.Collections.Generic;
using AutoMapper;
using CulinarySite.Bll.Interfaces;
using CulinarySite.Common.Dtos.Subscriber;
using CulinarySite.Dal.Interfaces;
using CulinarySite.Domain.Entities;
using System.Linq;

namespace CulinarySite.Bll.Services
{
    public class SubscriberService : ISubscriberService
    {
        private readonly IReadOnlyGenericRepository<Subscriber> _subscriberReadOnlyRepository;
        private readonly IWriteGenericRepository<Subscriber> _subscriberWriteRepository;
        private readonly IMapper _mapper;
        public SubscriberService(
            IReadOnlyGenericRepository<Subscriber> subscriberReadOnlyRepository,
            IWriteGenericRepository<Subscriber> subscriberWriteRepository,
            IMapper mapper)
        {
            _subscriberReadOnlyRepository = subscriberReadOnlyRepository;
            _subscriberWriteRepository = subscriberWriteRepository;
            _mapper = mapper;
        }

        public void CreateSubscriber(CreateSubscriberDto createSubscriberDto)
        {
            Subscriber subscriber = _mapper.Map<Subscriber>(createSubscriberDto);

            _subscriberWriteRepository.Create(subscriber);
            _subscriberWriteRepository.Save();
        }

        public void UpdateSubscriber(UpdateSubscriberDto updateSubscriberDto)
        {
            Subscriber subscriber = _mapper.Map<Subscriber>(updateSubscriberDto);

            _subscriberWriteRepository.Update(subscriber);
            _subscriberWriteRepository.Save();
        }

        public void DeleteSubscriber(int id)
        {
            _subscriberWriteRepository.Delete(id);
            _subscriberWriteRepository.Save();
        }

        public IEnumerable<SubscriberListDto> GetSubscriberList(bool withRelated)
        {
            IEnumerable<Subscriber> subscribers;
            IEnumerable<SubscriberListDto> subscriberListDtos ;

            if (withRelated)
            {
                subscribers = _subscriberReadOnlyRepository.GetItemListWithInclude(x => x.Comments);
                subscriberListDtos = subscribers.Select(x => _mapper.Map<SubscriberListDto>(x));
               
                return subscriberListDtos;
            }

            subscribers = _subscriberReadOnlyRepository.GetItemList();
            subscriberListDtos = subscribers.Select(x => _mapper.Map<SubscriberListDto>(x));

            return subscriberListDtos;
        }

        public SubscriberDetailDto GetSubscriber(int id, bool withRelated)
        {
            var subscriber = new Subscriber();
            var subscriberDetailDto = new SubscriberDetailDto();

            if (withRelated)
            {
                subscriber = _subscriberReadOnlyRepository.GetItemWithInclude(
                    x => x.Id == id,
                    x => x.Comments);

                subscriberDetailDto = _mapper.Map<SubscriberDetailDto>(subscriber);

                return subscriberDetailDto;
            }

            subscriber = _subscriberReadOnlyRepository.GetItem(id);

            subscriberDetailDto = _mapper.Map<SubscriberDetailDto>(subscriber);

            return subscriberDetailDto;
        }

    }
}
